client/rpc-spec-v2/src/chain_head/chain_head.rs

```diff
@@ -0,0 +1,785 @@
+			Initialized, NetworkConfig, NewBlock, RuntimeEvent, RuntimeVersionEvent,
+		},
+		subscription::{SubscriptionHandle, SubscriptionManagement, SubscriptionManagementError},
+	},
+	SubscriptionTaskExecutor,
+};
+use codec::Encode;
+use futures::{
+	channel::oneshot,
+	future::FutureExt,
+	stream::{self, Stream, StreamExt},
+};
+use futures_util::future::Either;
+use jsonrpsee::{
+	core::{async_trait, RpcResult},
+	types::{SubscriptionEmptyError, SubscriptionId, SubscriptionResult},
+	SubscriptionSink,
+};
+use log::{debug, error};
+use sc_client_api::{
+	Backend, BlockBackend, BlockImportNotification, BlockchainEvents, CallExecutor, ChildInfo,
+	ExecutorProvider, FinalityNotification, StorageKey, StorageProvider,
+};
+use serde::Serialize;
+use sp_api::CallApiAt;
+use sp_blockchain::{
+	Backend as BlockChainBackend, Error as BlockChainError, HeaderBackend, HeaderMetadata,
+};
+use sp_core::{hexdisplay::HexDisplay, storage::well_known_keys, Bytes};
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use std::{marker::PhantomData, sync::Arc};
+
+/// An API for chain head RPC calls.
+pub struct ChainHead<BE, Block: BlockT, Client> {
+	/// Substrate client.
+	client: Arc<Client>,
+	/// Backend of the chain.
+	backend: Arc<BE>,
+	/// Executor to spawn subscriptions.
+	executor: SubscriptionTaskExecutor,
+	/// Keep track of the pinned blocks for each subscription.
+	subscriptions: Arc<SubscriptionManagement<Block>>,
+	/// The hexadecimal encoded hash of the genesis block.
+	genesis_hash: String,
+	/// The maximum number of pinned blocks allowed per connection.
+	max_pinned_blocks: usize,
+	/// Phantom member to pin the block type.
+	_phantom: PhantomData<Block>,
+}
+
+impl<BE, Block: BlockT, Client> ChainHead<BE, Block, Client> {
+	/// Create a new [`ChainHead`].
+	pub fn new<GenesisHash: AsRef<[u8]>>(
+		client: Arc<Client>,
+		backend: Arc<BE>,
+		executor: SubscriptionTaskExecutor,
+		genesis_hash: GenesisHash,
+		max_pinned_blocks: usize,
+	) -> Self {
+		let genesis_hash = format!("0x{:?}", HexDisplay::from(&genesis_hash.as_ref()));
+
+		Self {
+			client,
+			backend,
+			executor,
+			subscriptions: Arc::new(SubscriptionManagement::new()),
+			genesis_hash,
+			max_pinned_blocks,
+			_phantom: PhantomData,
+		}
+	}
+
+	/// Accept the subscription and return the subscription ID on success.
+	///
+	/// Also keep track of the subscription ID internally.
+	fn accept_subscription(
+		&self,
+		sink: &mut SubscriptionSink,
+	) -> Result<String, SubscriptionEmptyError> {
+		// The subscription must be accepted before it can provide a valid subscription ID.
+		sink.accept()?;
+
+		let Some(sub_id) = sink.subscription_id() else {
+			// This can only happen if the subscription was not accepted.
+			return Err(SubscriptionEmptyError)
+		};
+
+		// Get the string representation for the subscription.
+		let sub_id = match sub_id {
+			SubscriptionId::Num(num) => num.to_string(),
```

Comment:
> Return value: String containing an opaque value representing the subscription.

It's technically possible to provide a `custom SubscriptionId` such number but perhaps we could enforce substrate to only accept something that serializes to a String.

I think it's better to reject the subscription if the subscription ID is a number feels like bug for someone to have use another custom subscription ID generator but it's converted to a String under the hood.

This is a edge-case and we could deal with it in another PR.